hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java

```diff
@@ -1571,11 +1560,33 @@ public List<BlockGroup> getPendingDeletionKeys(final int keyCount,
             }
             currentCount++;
           }
-          keyBlocksList.addAll(blockGroupList);
+
+          List<OmKeyInfo> notReclaimableKeyInfoList =
+              notReclaimableKeyInfo.getOmKeyInfoList();
+
+          // If all the versions are not reclaimable, then do nothing.
+          if (notReclaimableKeyInfoList.size() > 0 &&
+              notReclaimableKeyInfoList.size() !=
+                  infoList.getOmKeyInfoList().size()) {
+            keysToModify.put(kv.getKey(), notReclaimableKeyInfo);
+          }
+
+          if (notReclaimableKeyInfoList.size() !=
+              infoList.getOmKeyInfoList().size()) {
+            keyBlocksList.addAll(blockGroupList);
+          }
         }
       }
     }
-    return keyBlocksList;
+    return new PendingKeysDeletion(keyBlocksList, keysToModify);
+  }
+
+  private boolean versionExistsInPreviousSnapshot(OmKeyInfo omKeyInfo,
+      OmKeyInfo info, RepeatedOmKeyInfo delOmKeyInfo) {
+    return (omKeyInfo != null &&
```

Comment:
Need handle for HSync, where object will be same but blocks are different as append mechanism

Reply:
I'm confused, In hsync the keys will be `OpenKeyTable` until it is committed right? We here only care about those keys in `keyTable` and `deletedTable`. Please correct me if I'm wrong.

Reply:
@aswinshakil 
In Hsync, 
- Key will be present in openKeyTable and keyTable both
- when commit happens, the blocks at that moment will be flushed to keyTable, so that is available for read
- then further new blocks can be added in openKeyTable and flush to commit table

So there is a possibility,
- snapshot can have "key" in keyTable which is having partial data
- So both key needs to be concluded as same key only (no need compare blocks)

One way to solve is, previous key need to check in metada, if Hsync flag is present or not. If present, then no need compare blocks.

Reply:
Thanks for the explanation. I will update the PR to have a hsync check.